app/ts/Workspace/WorkspaceContext.tsx

```diff
@@ -86,23 +111,153 @@ export function WorkspaceContextProviderComponent({
+      {},
+      _.cloneDeep(oldWorkspace),
+      {problem: {alternatives: {[alternative.id]: newAlternative}}}
+    );
+    setWorkspace(buildWorkspace(oldWorkspaceToSend, workspaceId, ordering));
+    sendOldWorkspace(oldWorkspaceToSend);
+  }
+
+  function editCriterion(newCriterion: ICriterion): void {
+    const oldWorkspaceToSend: IOldWorkspace = _.merge(
+      {},
+      _.cloneDeep(oldWorkspace),
+      {
+        problem: {
+          criteria: {
+            [newCriterion.id]: transformCriterionToOldCriterion(newCriterion)
+          }
+        }
+      }
+    );
+    setWorkspace(buildWorkspace(oldWorkspaceToSend, workspaceId, ordering));
+    sendOldWorkspace(oldWorkspaceToSend);
+  }
+
+  function sendOldWorkspace(oldWorkspaceToSend: IOldWorkspace) {
+    Axios.post(`/workspaces/${workspaceId}`, oldWorkspaceToSend)
+      .catch(errorCallback)
+      .then(() => {
+        window.location.reload();
+      }); // FIXME: needed to update the angular scope
+  }
+
   function errorCallback(error: OurError) {
     setError(error);
   }
 
+  function swapAlternatives(
+    alternative1Id: string,
+    alternative2Id: string
+  ): void {
+    const newAlternatives = swapItems(
+      alternative1Id,
+      alternative2Id,
+      workspace.alternatives
+    );
+    setWorkspace(
+      _.merge({}, _.cloneDeep(workspace), {alternatives: newAlternatives})
+    );
+    const newOrdering: IOrdering = {
+      alternatives: _.map(newAlternatives, 'id'),
+      criteria: _.map(workspace.criteria, 'id'),
+      dataSources: _.flatMap(
+        workspace.criteria,
+        (criterion: ICriterion): string[] => _.map(criterion.dataSources, 'id')
+      )
+    };
+    Axios.put(`/workspaces/${workspaceId}/ordering`, newOrdering).catch(
+      errorCallback
+    );
+  }
+
+  function swapCriteria(criterion1Id: string, criterion2Id: string): void {
+    const newCriteria = swapItems(
+      criterion1Id,
+      criterion2Id,
+      workspace.criteria
+    );
+    setWorkspace(_.merge({}, _.cloneDeep(workspace), {criteria: newCriteria}));
+    const newOrdering: IOrdering = {
+      alternatives: _.map(workspace.alternatives, 'id'),
+      criteria: _.map(newCriteria, 'id'),
+      dataSources: _.flatMap(
+        workspace.criteria,
+        (criterion: ICriterion): string[] => _.map(criterion.dataSources, 'id')
+      )
+    };
+    Axios.put(`/workspaces/${workspaceId}/ordering`, newOrdering).catch(
+      errorCallback
+    );
+  }
+
+  function swapDataSources(
+    criterionId: string,
+    dataSource1Id: string,
+    dataSource2Id: string
+  ): void {
+    const criterionIndex = _.findIndex(workspace.criteria, ['id', criterionId]);
+    const criterion = workspace.criteria[criterionIndex];
+    const newCriterion = {
+      ...criterion,
+      dataSources: swapItems(
+        dataSource1Id,
+        dataSource2Id,
+        criterion.dataSources
+      )
+    };
+    let newCriteria = _.cloneDeep(workspace.criteria);
```

Comment:
I don't even think clonedeep is necessary any more when we're just consting things around
```
const newCriteria = {
   ..._.cloneDeep(workspace.criteria),
   [criterionIndex]: newCriterion
}
```

Reply:
Tried this, but it doesn't work for arrays (at least for me)

Reply:
oh, arrays, yes, then it won't work.